chromadb/api/shared_system_client.py

```diff
@@ -94,3 +114,62 @@ def _system(self) -> System:
     def _submit_client_start_event(self) -> None:
         telemetry_client = self._system.instance(ProductTelemetryClient)
         telemetry_client.capture(ClientStartEvent())
+
+    @staticmethod
+    def get_chroma_cloud_api_key_from_clients() -> Optional[str]:
+        """
+        Try to extract api key from existing client instances by checking httpx session headers.
+
+        Requirements to pull api key:
+        - must be a FastAPI or AsyncFastAPI instance (ignore RustBindingsAPI and SegmentAPI)
+        - must have "api.trychroma.com" or "gcp.trychroma.com" in the _api_url (ignore local/self-hosted instances)
+        - must have "x-chroma-token" or "X-Chroma-Token" in the headers
+
+        Returns:
+            The first api key found, or None if no client instances have api keys set.
+        """
+        FastAPI = _get_fastapi_class()
+        AsyncFastAPI = _get_async_fastapi_class()
+
+        # track all api keys from eligible Chroma Cloud clients
+        api_keys: list[str] = []
+        systems_snapshot = list(SharedSystemClient._identifier_to_system.values())
+        for system in systems_snapshot:
+            try:
+                server_api = system.instance(ServerAPI)
+
+                # get headers based on client type
+                headers: Mapping[str, Any]
+                if isinstance(server_api, FastAPI):
+                    headers = server_api._session.headers
+                elif isinstance(server_api, AsyncFastAPI):
+                    headers = server_api._settings.chroma_server_headers or {}
+                else:
+                    # RustBindingsAPI and SegmentAPI don't have HTTP headers
+                    continue
+
+                # Only pull api key if the url contains the chroma cloud url
+                if (
+                    "api.trychroma.com" not in server_api._api_url
+                    and "gcp.trychroma.com" not in server_api._api_url
+                ):
+                    continue
+
+                # Pull api key from the chroma token header
+                api_key = headers.get("X-Chroma-Token") or headers.get("x-chroma-token")
```

Comment:
If you want this to be case insensitive why not just enumerate the headers, down case and extract the match? This handles all cases as opposed to these two.